### ISSUE TITLE
XSLT error when no allowances/charges are present in the document

### ISSUE THREAD

**Issue by congory** [ZUGFeRD/mustangproject#496]
When there are no allowances or charges present in the document the XSLT triggers the following errors:

<validation filename="factur-x.xml" datetime="2024-09-30 20:42:02">
  <xml>
    <info>
      <version>2</version>
      <profile>urn:cen.eu:en16931:2017#conformant#urn:factur-x.eu:1p0:extended</profile>
      <validator version="2.14.0"/>
      <rules>
        <fired>110</fired>
        <failed>2</failed>
      </rules>
      <duration unit="ms">2168</duration>
    </info>
    <messages>
      <error type="4" location="/*:CrossIndustryInvoice[namespace-uri()='urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100'][1]/*:SupplyChainTradeTransaction[namespace-uri()='urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100'][1]/*:ApplicableHeaderTradeSettlement[namespace-uri()='urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100'][1]/*:SpecifiedTradeSettlementHeaderMonetarySummation[namespace-uri()='urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100'][1]" criterion="for $calculatedAmount in xs:decimal(round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='false']/ram:ActualAmount)* 10 * 10 ) div 100), $totalAmount in xs:decimal(ram:AllowanceTotalAmount), $nbAllowanceItems in xs:decimal(count(../ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='false'])), $tolerance in xs:decimal(0.01), $maxTolerance in $nbAllowanceItems * $tolerance, $diff in xs:decimal($totalAmount - $calculatedAmount), $abs in xs:decimal(abs($diff)) return $abs le $maxTolerance or (not(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='false']) and not (ram:AllowanceTotalAmount))">[BR-FXEXT-CO-11]-Absolute Value of (Sum of allowances on document level (BT-107) - Σ Document level allowance amounts (BT-92))&lt;= 0,01 * Number of Document level allowance amounts (BT-92). [ID FX-SCH-A-000304] from /xslt/ZF_230/FACTUR-X_EXTENDED.xslt)</error>  
      <error type="4" location="/*:CrossIndustryInvoice[namespace-uri()='urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100'][1]/*:SupplyChainTradeTransaction[namespace-uri()='urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100'][1]/*:ApplicableHeaderTradeSettlement[namespace-uri()='urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100'][1]/*:SpecifiedTradeSettlementHeaderMonetarySummation[namespace-uri()='urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100'][1]" criterion="for $calculatedAmount in xs:decimal(xs:decimal(round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='true']/ram:ActualAmount)* 100 ) div 100)+ xs:decimal(round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedLogisticsServiceCharge/ram:AppliedAmount)* 10 * 10 ) div 100)), $totalAmount in xs:decimal(ram:ChargeTotalAmount), $nbChargeItems in xs:decimal(count(../ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='true']) + count(../ram:SpecifiedLogisticsServiceCharge)), $tolerance in xs:decimal(0.01), $maxTolerance in $nbChargeItems * $tolerance, $diff in xs:decimal($totalAmount - $calculatedAmount), $abs in xs:decimal(abs($diff)) return $abs le $maxTolerance or (not(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='true'])and not (ram:ChargeTotalAmount))">[BR-FXEXT-CO-12]-Absolute Value of (Sum of charges on document level (BT-108) - Σ Document level charge amounts (BT-99) - Σ Logistics Service fee amounts (BT-x-272))&lt;= 0,01 * (Number of Document level charge amounts (BT-99) + Number of Logistics Service fee amounts (BT-X-272)). [ID FX-SCH-A-000305] from /xslt/ZF_230/FACTUR-X_EXTENDED.xslt)</error> 
    </messages>
    <summary status="invalid"/>
  </xml>
  <messages></messages>
  <summary status="invalid"/>
</validation>

**Comment by jstaerk** [comment]
could you please add a concrete xml sample?
If you wanted to remove personal data you could e.g. try [cii anonymizer](https://github.com/ZUGFeRD/einvoice-anonymizer) just watch out and copy the file before anonymizing, it overwrites the original.

**Comment by congory** [comment]
Hi Jochen, here you go. Should be fully anonymized.

<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<rsm:CrossIndustryInvoice xmlns:ram="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100" xmlns:rsm="urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100" xmlns:udt="urn:un:unece:uncefact:data:standard:UnqualifiedDataType:100" xmlns:qdt="urn:un:unece:uncefact:data:standard:QualifiedDataType:100" xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <rsm:ExchangedDocumentContext>
        <ram:BusinessProcessSpecifiedDocumentContextParameter>
            <ram:ID>Software Consulting and Development Services</ram:ID>
        </ram:BusinessProcessSpecifiedDocumentContextParameter>
        <ram:GuidelineSpecifiedDocumentContextParameter>
            <ram:ID>urn:cen.eu:en16931:2017#conformant#urn:factur-x.eu:1p0:extended</ram:ID>
        </ram:GuidelineSpecifiedDocumentContextParameter>
    </rsm:ExchangedDocumentContext>
    <rsm:ExchangedDocument>
        <ram:ID>DO NOT PAY - TEST INVOICE</ram:ID>
        <ram:TypeCode>380</ram:TypeCode>
        <ram:IssueDateTime>
            <udt:DateTimeString format="102">20241007</udt:DateTimeString>
        </ram:IssueDateTime>
        <ram:IncludedNote>
            <ram:Content>DO NOT PAY - TEST INVOICE</ram:Content>
            <ram:SubjectCode>AAI</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content> 
Software Consulting Mustermann d.d. 
Not existent Street 32 
10000 Zagreb 
Hrvatska / Croatia / Kroatien 
Registar / Registry / Register: Trgovački sud u Zagrebu, ABC 123456 
Temeljni kapital od 20.000 kn uplaćen je u cijelosti / Registered capital of 20,000 kn fully paid / Stammkapital von 20.000 kn vollständig eingezahlt 
Član uprave / Board member / Vorstand: Mustermann</ram:Content>
            <ram:SubjectCode>ABL</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content>
REVERSE CHARGE: 
Prema čl. 44. EU Direktive vijeća 2006/112/EZ Porez na dodanu vrijednost treba prijaviti i platiti od strane primatelja usluge. / 
According to art. 44 of the EU council directive 2006/112/EC the value-added tax must be declared and paid by the service recipient. / 
Entsprechend Art. 44 der EU Ratsrichtlinie 2006/112/EG ist die Umsatzsteuer vom Leistungsempfänger anzumelden und abzuführen.</ram:Content>
            <ram:SubjectCode>REG</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content>Vremenski žig i mjesto izdavanja računa / 
Timestamp and location of issue / 
Zeitstempel und Ort der Ausstellung: 
07.10.2024 13:27 HR-Zagreb</ram:Content>
            <ram:SubjectCode>REG</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content>Oznaka operatera na naplatnom uređaju / 
Cashier Id / 
Bediener Id: 
1</ram:Content>
            <ram:SubjectCode>REG</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content>Osoba odgovorna za izdavanje računa / 
Billing responsible person / 
Verantwortlicher Rechnungsaussteller: 
Kassenmann</ram:Content>
            <ram:SubjectCode>REG</ram:SubjectCode>
        </ram:IncludedNote>
        <ram:IncludedNote>
            <ram:Content>Seller full bank details / 
 Bank(a): Zagrebačka banka (UniCredit Group) / 
 IBAN: HR12 3456 7890 1011 1213 1 / 
 BIC/SWIFT: ZABAHR2XXXX</ram:Content>
            <ram:SubjectCode>REG</ram:SubjectCode>
        </ram:IncludedNote>
    </rsm:ExchangedDocument>
    <rsm:SupplyChainTradeTransaction>
        <ram:IncludedSupplyChainTradeLineItem>
            <ram:AssociatedDocumentLineDocument>
                <ram:LineID>1</ram:LineID>
            </ram:AssociatedDocumentLineDocument>
            <ram:SpecifiedTradeProduct>
                <ram:Name>INDIVIDUAL_SW_DEVELOPMENT</ram:Name>
                <ram:Description>Izrada individualnog softvera / Development individual software / Entwicklung Individual Software</ram:Description>
            </ram:SpecifiedTradeProduct>
            <ram:SpecifiedLineTradeAgreement>
                <ram:GrossPriceProductTradePrice>
                    <ram:ChargeAmount>5.00</ram:ChargeAmount>
                </ram:GrossPriceProductTradePrice>
                <ram:NetPriceProductTradePrice>
                    <ram:ChargeAmount>5.00</ram:ChargeAmount>
                </ram:NetPriceProductTradePrice>
            </ram:SpecifiedLineTradeAgreement>
            <ram:SpecifiedLineTradeDelivery>
                <ram:BilledQuantity unitCode="HUR">160.00</ram:BilledQuantity>
            </ram:SpecifiedLineTradeDelivery>
            <ram:SpecifiedLineTradeSettlement>
                <ram:ApplicableTradeTax>
                    <ram:TypeCode>VAT</ram:TypeCode>
                    <ram:ExemptionReason>Obračun PDV-a prema prijenosu porezne obveze / 
 VAT calculated under reverse charge mechanism / 
 USt. gemäß Reverse-Charge-Verfahren</ram:ExemptionReason>
                    <ram:CategoryCode>AE</ram:CategoryCode>
                    <ram:ExemptionReasonCode>VATEX-EU-AE</ram:ExemptionReasonCode>
                    <ram:RateApplicablePercent>0</ram:RateApplicablePercent>
                </ram:ApplicableTradeTax>
                <ram:BillingSpecifiedPeriod>
                    <ram:StartDateTime>
                        <udt:DateTimeString format="102">20240901</udt:DateTimeString>
                    </ram:StartDateTime>
                    <ram:EndDateTime>
                        <udt:DateTimeString format="102">20240930</udt:DateTimeString>
                    </ram:EndDateTime>
                </ram:BillingSpecifiedPeriod>
                <ram:SpecifiedTradeSettlementLineMonetarySummation>
                    <ram:LineTotalAmount>800.00</ram:LineTotalAmount>
                </ram:SpecifiedTradeSettlementLineMonetarySummation>
            </ram:SpecifiedLineTradeSettlement>
        </ram:IncludedSupplyChainTradeLineItem>
        <ram:ApplicableHeaderTradeAgreement>
            <ram:BuyerReference>Ugovor/Contract/Vertrag Mustervertrag_XYZ, 26.09.2023</ram:BuyerReference>
            <ram:SellerTradeParty>
                <ram:Name>Software Consulting Mustermann d.d.</ram:Name>
                <ram:DefinedTradeContact>
                    <ram:PersonName>Mustermann</ram:PersonName>
                    <ram:TelephoneUniversalCommunication>
                        <ram:CompleteNumber>+385 11111111111</ram:CompleteNumber>
                    </ram:TelephoneUniversalCommunication>
                    <ram:EmailURIUniversalCommunication>
                        <ram:URIID>not_there_Mustermann@gmail.com</ram:URIID>
                    </ram:EmailURIUniversalCommunication>
                </ram:DefinedTradeContact>
                <ram:PostalTradeAddress>
                    <ram:PostcodeCode>10000</ram:PostcodeCode>
                    <ram:LineOne>Not existent Street 32</ram:LineOne>
                    <ram:CityName>Zagreb</ram:CityName>
                    <ram:CountryID>HR</ram:CountryID>
                </ram:PostalTradeAddress>
                <ram:URIUniversalCommunication>
                    <ram:URIID schemeID="EM">not_there_Mustermann@gmail.com</ram:URIID>
                </ram:URIUniversalCommunication>
                <ram:SpecifiedTaxRegistration>
                    <ram:ID schemeID="VA">HR123456788888</ram:ID>
                </ram:SpecifiedTaxRegistration>
                <ram:SpecifiedTaxRegistration>
                    <ram:ID schemeID="FC">123456788888</ram:ID>
                </ram:SpecifiedTaxRegistration>
            </ram:SellerTradeParty>
            <ram:BuyerTradeParty>
                <ram:Name>Musterkunde Gibtsnicht AG</ram:Name>
                <ram:PostalTradeAddress>
                    <ram:PostcodeCode>12345</ram:PostcodeCode>
                    <ram:LineOne>Musterkunde-Straße 8</ram:LineOne>
                    <ram:CityName>Köln</ram:CityName>
                    <ram:CountryID>DE</ram:CountryID>
                </ram:PostalTradeAddress>
                <ram:URIUniversalCommunication>
                    <ram:URIID schemeID="EM">abrechnung@musterkunde.mu</ram:URIID>
                </ram:URIUniversalCommunication>
                <ram:SpecifiedTaxRegistration>
                    <ram:ID schemeID="VA">DE1234567890000</ram:ID>
                </ram:SpecifiedTaxRegistration>
            </ram:BuyerTradeParty>
            <ram:BuyerOrderReferencedDocument>
                <ram:IssuerAssignedID>Ugovor/Contract/Vertrag Mustervertrag_XYZ, 26.09.2023</ram:IssuerAssignedID>
            </ram:BuyerOrderReferencedDocument>
        </ram:ApplicableHeaderTradeAgreement>
        <ram:ApplicableHeaderTradeDelivery>
            <ram:ActualDeliverySupplyChainEvent>
                <ram:OccurrenceDateTime>
                    <udt:DateTimeString format="102">20240930</udt:DateTimeString>
                </ram:OccurrenceDateTime>
            </ram:ActualDeliverySupplyChainEvent>
        </ram:ApplicableHeaderTradeDelivery>
        <ram:ApplicableHeaderTradeSettlement>
            <ram:PaymentReference>DO NOT PAY - TEST INVOICE</ram:PaymentReference>
            <ram:InvoiceCurrencyCode>EUR</ram:InvoiceCurrencyCode>
            <ram:SpecifiedTradeSettlementPaymentMeans>
                <ram:TypeCode>42</ram:TypeCode>
                <ram:Information>PAYMENT_TO_BANK_ACCOUNT</ram:Information>
                <ram:PayeePartyCreditorFinancialAccount>
                    <ram:IBANID>HR12 3456 7890 1011 1213 1</ram:IBANID>
                    <ram:AccountName>Software Consulting Mustermann d.d.</ram:AccountName>
                </ram:PayeePartyCreditorFinancialAccount>
            </ram:SpecifiedTradeSettlementPaymentMeans>
            <ram:ApplicableTradeTax>
                <ram:CalculatedAmount>0.00</ram:CalculatedAmount>
                <ram:TypeCode>VAT</ram:TypeCode>
                <ram:ExemptionReason>Obračun PDV-a prema prijenosu porezne obveze / 
 VAT calculated under reverse charge mechanism / 
 USt. gemäß Reverse-Charge-Verfahren</ram:ExemptionReason>
                <ram:BasisAmount>800.00</ram:BasisAmount>
                <ram:CategoryCode>AE</ram:CategoryCode>
                <ram:ExemptionReasonCode>VATEX-EU-AE</ram:ExemptionReasonCode>
                <ram:RateApplicablePercent>0</ram:RateApplicablePercent>
            </ram:ApplicableTradeTax>
            <ram:BillingSpecifiedPeriod>
                <ram:StartDateTime>
                    <udt:DateTimeString format="102">20240901</udt:DateTimeString>
                </ram:StartDateTime>
                <ram:EndDateTime>
                    <udt:DateTimeString format="102">20240930</udt:DateTimeString>
                </ram:EndDateTime>
            </ram:BillingSpecifiedPeriod>
            <ram:SpecifiedTradePaymentTerms>
                <ram:DueDateDateTime>
                    <udt:DateTimeString format="102">20241106</udt:DateTimeString>
                </ram:DueDateDateTime>
            </ram:SpecifiedTradePaymentTerms>
            <ram:SpecifiedTradeSettlementHeaderMonetarySummation>
                <ram:LineTotalAmount>800.00</ram:LineTotalAmount>
                <ram:TaxBasisTotalAmount>800.00</ram:TaxBasisTotalAmount>
                <ram:TaxTotalAmount currencyID="EUR">0.00</ram:TaxTotalAmount>
                <ram:GrandTotalAmount>800.00</ram:GrandTotalAmount>
                <ram:TotalPrepaidAmount>0.00</ram:TotalPrepaidAmount>
                <ram:DuePayableAmount>800.00</ram:DuePayableAmount>
            </ram:SpecifiedTradeSettlementHeaderMonetarySummation>
        </ram:ApplicableHeaderTradeSettlement>
    </rsm:SupplyChainTradeTransaction>
</rsm:CrossIndustryInvoice>

**Comment by SemikolonDEV** [comment]
The error can also be reproduced with the sample file “EXTENDED_ProjectCompletionInvoice.pdf” and “EXTENDED_InternalCommunityDeliveryMultipleOrders.pdf”.

![image](https://github.com/user-attachments/assets/3014c99e-62e1-4e6c-aef4-786450ea5750)

I think it is a bug in the testing of BR-FXEXT-CO-11 and BR-FXEXT-CO-12 rules.

When you add, `<ram:ChargeTotalAmount>0</ram:ChargeTotalAmount >` and  		`<ram:AllowanceTotalAmount>0</ram:AllowanceTotalAmount>` then there is no Error.

Example:
```xml
	<ram:SpecifiedTradeSettlementHeaderMonetarySummation>
		<ram:LineTotalAmount>2000.00</ram:LineTotalAmount>
		<ram:ChargeTotalAmount>0</ram:ChargeTotalAmount >
		<ram:AllowanceTotalAmount>0</ram:AllowanceTotalAmount>
		<ram:TaxBasisTotalAmount>2000.00</ram:TaxBasisTotalAmount>
		<ram:TaxTotalAmount currencyID="EUR">0.00</ram:TaxTotalAmount>
		<ram:GrandTotalAmount>2000.00</ram:GrandTotalAmount>
		<ram:DuePayableAmount>2000.00</ram:DuePayableAmount>
	</ram:SpecifiedTradeSettlementHeaderMonetarySummation>
```

**Comment by AlienAtSystem** [comment]
I also just encountered this error today, and it needs to be fixed on code side: According to the Faktur-X documentation, the lines ChargeTotalAmount and AllowanceTotalAmount are mandatory **only** if there are SpecifiedTradeAllowanceCharge <Document Level Allowances> or SpecifiedTradeAllowanceCharge <Document Level Charges> entries that need to be summed up.
That means, instead of just blindly checking BR-CO-11 and BR-CO-12, it first needs to check whether there *are* entries that fall under these rules, then whether the required TotalAmount entry is present, and only then whether the sums add up.

**Comment by Malove98** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by jstaerk** [comment]
thats fixed in zugferd 2.3.2 (wasnt that even earlier?) and mustangproject 2.15.0 at the latest
